motr/client.c

```diff
@@ -830,6 +830,66 @@ M0_INTERNAL int m0_op_init(struct m0_op *op,
 	return M0_RC(0);
 }
 
+M0_INTERNAL int m0_calculate_md5_inc_digest(
+		struct m0_md5_inc_digest_pi *pi,
+		struct m0_pi_seed *seed,
+		struct m0_bufvec *bvec,
+		enum m0_pi_calc_flag flag,
+		unsigned char *curr_digest,
+		unsigned char *checksum_without_seed)
+{
+	M0_ENTRY();
+
+	M0_PRE(pi != NULL);
+	M0_PRE(curr_digest != NULL);
+	M0_PRE(bvec != NULL);
+	M0_PRE(bvec->ov_vec.v_count != NULL);
+	M0_PRE(bvec->ov_buf != NULL);
+
+	/* Uncomment following code once motr has integration for MD5 */
+#if 0
+
+	/* This call is for first data unit, need to initialize prev_digest*/
+	if (flag & M0_PI_CALC_UNIT_ZERO) {
+		MD5_Init(&md5_digest_pi->prev_digest);
+	}
+
+	/* memcpy, so that we do not change the prev_digest */
+	memcpy(curr_digest, &md5_digest_pi->prev_digest, sizeof(MD5_CTX));
+
+	/* get the curr digest i by updating it*/
+	for (i = 0; i < bvec->ov_vec.v_nr; i++) {
+		MD5_Update(curr_digest, bvec->ov_buf[i], bvec->ov_vec.v_count[i]);
+	}
+
+	/* if caller want checksum without seed and with seed, in this case
+	 * 'checksum_without_seed' will be used to send wihout seed checksum
+	 */
+	if (checksum_without_seed) {
+		MD5_Final(checksum_without_seed, curr_digest);
+	}
+
+	/* if seed is passed, memcpy and update the digest calculated so far.
+	 * calculate checksum with seed, set the pi_value with seeded checksum.
+	 * if sedd is NULL, populate pi_value with unseeded checksum.
+	 * NOTE: curr_digest will always have digest without seed.
+	 */
+	if (seed) {
+		unsigned char seed_digest[sizeof(MD5_CTX)];
+		memcpy(&seed_digest, curr_digest, sizeof(MD5_CTX));
+
+		MD5_Update(&seed_digest, &seed, sizeof(struct m0_pi_seed));
+		MD5_Final(md5_digest_pi->pi_value, &seed_digest);
+	}
+	else {
```

Comment:
if seed is not set calling MD5_final() is not needed since
for every 16K buffers we will be calling this api, for that only MD5_Update() to be called
So
1) with flag M0_PI_CALC_UNIT_ZERO
        MD5_Init()
2) Then for each 16k buffers motr api will call MD5_Update()

3) After 64th buffer (1MB motr unit size) will send seed hence it will be
       MD5_Update with seed and then MD5_Final()

Reply:
We are having only one for loop and it does MD5_update(), not MD5_final()

I believe we will get multiple 16k buffers that is why we have m0_bufvec for the data.

I do not understand how  MD5_Final() will get called for each 16k buffers.

Is S3 going to send only one 16k buffer at a time ?

Reply:
Please see my comment at line#876

Reply:
On checking the code with Rajesh and Nilesh, it was seen that MD5_Final is modifying the digest and it may cause some issue in computing full checksum and does lot of computation which we can avoid. So we can add flag to skip this setup.

Reply:
Also the curr_digest should be non finalized one i.e. it should never be part of MD5_Final call